emcc.py

```diff
@@ -1438,16 +1438,17 @@ def check(input_file):
       if any(s.startswith('MEM_INIT_METHOD=') for s in settings_changes):
         exit_with_error('MEM_INIT_METHOD is not supported in wasm. Memory will be embedded in the wasm binary if threads are not used, and included in a separate file if threads are used.')
       if shared.Settings.WASM2JS:
-        # in wasm2js, keep the mem init in the wasm itself if we can (no pthreads), and if the options
-        # wouldn't tell a js build to use a separate mem init file
-        shared.Settings.MEM_INIT_IN_WASM = not shared.Settings.USE_PTHREADS and not options.memory_init_file
+        # wasm2js does not support passive segments or atomics
+        if shared.Settings.USE_PTHREADS:
+          exit_with_error('WASM2JS does not yet support pthreads')
+        # in wasm2js, keep the mem init in the wasm itself if we can and if the
+        # options wouldn't tell a js build to use a separate mem init file
+        shared.Settings.MEM_INIT_IN_WASM = not options.memory_init_file
       else:
-        # wasm normally includes the mem init in the wasm binary. exceptions are pthreads, where we need
-        # to split it out until we have full bulk memory support, and wasm2js, which behaves more like js.
+        # wasm includes the mem init in the wasm binary. The exception is
+        # wasm2js, which behaves more like js.
         options.memory_init_file = True
-        # we will include the mem init data in the wasm, when we don't need the
-        # mem init file to be loadable by itself
-        shared.Settings.MEM_INIT_IN_WASM = not shared.Settings.USE_PTHREADS
+        shared.Settings.MEM_INIT_IN_WASM = True if shared.Settings.WASM_BACKEND else not shared.Settings.USE_PTHREADS
```

Comment:
This is still confusing. I guess it will get simpler once fastcomp is gone, but why should `options.memory_init_file` not be the inverse of `shared.Settings.MEM_INIT_IN_WASM`?

Reply:
This is a great question and I don’t know the answer. @kripken did tell me it was complicated, though.

Reply:
I tried to simplify this but ran into fastcomp issues with how those are used (in bad ways). it seems more efficient to defer a cleanup til after fastcomp is removed.

Reply:
SGTM. This logic was confusing when I first added wasm backend support originally and I made it even worse then.